src/lib/cfa/sudoers.rb

```diff
@@ -0,0 +1,237 @@
+    # @option user_spec [String] "tag" the tag associated to the command (e.g. NOPASSWD)
+    def add_user(user_spec)
+      user = AugeasTree.new
+      host_group = AugeasTree.new
+
+      host_group.add("host", user_spec["host"])
+
+      commands = user_spec["commands"].reduce([]) do |commands, command|
+        augeas_command = AugeasTree.new
+        host_group.add("command[]", AugeasTreeValue.new(augeas_command, command))
+        commands << augeas_command
+      end
+
+      # For the time being, "runas_user" and "tag" options are added only to the first command,
+      # which means that they apply to all commands.
+      run_as = user_spec["run_as"].to_s.strip
+      tag = user_spec["tag"].to_s.strip
+      commands[0].add("runas_user", run_as) unless run_as.empty?
+      commands[0].add("tag", tag) unless tag.empty?
+
+      user.add("user", user_spec["user"])
+      user.add("host_group", host_group)
+
+      data.add("spec[]", user, placer)
+    end
+
+    private
+
+    attr_accessor :start_placer
+
+    # Return the current users collection
+    #
+    # @return [Array<Hash>]
+    def current_users
+      data.select(USERS_MATCHER)
+    end
+
+    COLLECTION_KEYS = ["spec", "command"].freeze
+    private_constant :COLLECTION_KEYS
+
+    # Fix collection names adding the missing "[]"
+    #
+    # When a collection has only one element, Augeas does not add "[]" suffix, reason why is
+    # necessary to fix at least those collextions that are going to be hitted for either, read or
+    # write.
+    #
+    # @param tree [AugeasTree,AugeasTreeValue]
+    def fix_collection_names(tree)
+      tree.data.each do |entry|
+        key, value = entry.values_at(:key, :value)
+
+        entry[:key] << "[]" if COLLECTION_KEYS.include?(key)
+        fix_collection_names(value) if value.is_a?(AugeasTree)
+        fix_collection_names(value.tree) if value.is_a?(AugeasTreeValue)
+      end
+    end
+
+    # Return the "next" placer where the user specification should be write
+    #
+    # When there are not users specifications yet, the start placer will be use. Otherwise, a new
+    # specification will be placed after the last one.
+    #
+    # @see #add_user
+    # @see #set_start_placer
+    #
+    # @return [CFA::Placer]
+    def placer
+      return start_placer if current_users.empty?
+
+      last_user = current_users.last
+
+      AfterPlacer.new(Matcher.new(key: last_user[:key], value_matcher: last_user[:value]))
+    end
+
+    # Set the the placer for the first user specification
+    #
+    # Depending on the defined users specifications, one of following criterias will be applied
+    #
+    #   * Generates the placer **after** the first one, or
+    #   * Look for the "#includedir .*sudoers.d.*" directive and generates the placer **before** it
+    #     if there are not users specifications yet.
+    #
+    # @see #placer
+    #
+    # @return [CFA::Placer]
+    def set_start_placer
```

Comment:
I am a bit confused here. I expect it will place user after previous users or before that include directive. but here it can also place it after first user. Which is a bit confusing for me.

Reply:
> I expect it will place user after previous users or before that include directive.

That's right.

> but here it can also place it after first user.

This is simply to respect the "original" position since yast-sudo will simply removes all the user and writes them again each time a new entry is added, deleted or modified. There is no way to edit an entry because it is not possible to identify it uniquely. What is more, for a same user we can have multiple user specifications in the sudoers file.

So, as said, this is simple for trying to respect the previous start position.

Do you think it does not worth it?

Reply:
yes, I think it makes sense such functionality, it is just not so much visible why it works. Reason why it works, is that matcher search in placer over `all_data` so it search even inside previously removed entries. So at least document how it is possible that it works as it is not obvious from first sight.